### PR TITLE
[improvement] Migrate to Java

### PR DESCRIPTION
**Pull request by ferozco**
## Before this PR
The majority of the project was written in groovy, with wide use of `afterEvaluate` to circumvent eager evaluation of task configuration.

## After this PR
I've converted to majority of the logic to Java and used as much modern gradle conventions as possible. I've verified usage against a 4.10 project and found that there where no break.

### CODE REVIEW

**Comment by iamdanfox** [comment]
I'm a little bit hesitant to do the groovy -> java rewrite and the behaviour change of always detectConstraints in the same PR... could we try to cut a release with just groovy -> java first? (I've done this for a couple of plugins so far and always found a couple of small unintended side effects of the rerite)

**Comment by ferozco** [comment]
@iamdanfox I totally understand the concern, I definitely want to do some thorough testing and an RC release before rolling this out more widely. I do want to make this easier to CR, but I don't think splitting it up would accomplish that. By consolidating the two changes, I was able to reduce the total change set by avoiding migrating a large amount of code and improving code reuse

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    private Property<String> serviceGroup;
+    private Property<String> serviceName;
+    private Property<String> podName;
+    private Property<ProductType> productType;
+    private MapProperty<String, Object> manifestExtensions;
+    private ListProperty<ProductDependency> productDependencies;
+    private SetProperty<ProductId> ignoredProductIds;
+
+    @Inject
+    public BaseDistributionExtension(Project project, ObjectFactory objectFactory) {
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.java]: I think it's more normal to just pass in `Project` and call `project.getObjects()` inside these constructors

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy [hunk 2]
```diff
@@ -30,27 +30,22 @@ class ProductDependency implements Serializable {
 
     private static final long serialVersionUID = 1L
 
+    @JsonProperty("product-group")
     String productGroup
-    String productName
 
-    @JsonIgnore
-    boolean detectConstraints
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy]: What does removing this mean for all the projects which currently set `detectConstraints = true`?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependency.groovy [hunk 3]
```diff
@@ -1,76 +0,0 @@
...
-import com.palantir.slspackaging.versions.SlsProductVersions
-import groovy.transform.AutoClone
-import groovy.transform.CompileStatic
-import groovy.transform.EqualsAndHashCode
-import groovy.transform.ToString
-
-@ToString
-@EqualsAndHashCode
-@AutoClone
-class RecommendedProductDependency {
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependency.groovy]: :+1: amusing that this was almost the same class as ProductDependency!

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/asset/AssetDistributionExtension.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.Map;
+import javax.inject.Inject;
+import org.gradle.api.Project;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.MapProperty;
+import org.gradle.api.provider.Provider;
+
+public class AssetDistributionExtension extends BaseDistributionExtension {
+
+    private MapProperty<String, String> assets;
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/asset/AssetDistributionExtension.java]: might as well make these private properties `final`!

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/asset/AssetDistributionPlugin.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            task.getArchiveBaseName().set(distributionExtension.getServiceName());
+            task.getArchiveVersion().set(project.provider(() -> project.getVersion().toString()));
+            task.getDestinationDirectory().set(new File(project.getBuildDir(), "distributions"));
+
+            task.from(new File(project.getProjectDir(), "deployment"));
+            task.from(new File(project.getBuildDir(), "deployment"));
+            task.dependsOn(manifest);
+        });
+
+        // HACKHACK after evaluate to configure task with all declared assets
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/asset/AssetDistributionPlugin.java]: Kinda sad this hack is necessary even though we're using providers everywhere... are you sure you can't do `task.from()` higher up and pass in a provider which is mapped from the distributionExtension.getAssets() ?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/PodDistributionExtension.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import org.gradle.api.Project;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.MapProperty;
+import org.gradle.api.provider.Provider;
+import org.gradle.util.ConfigureUtil;
+
+public class PodDistributionExtension extends BaseDistributionExtension {
+
+    private MapProperty<String, PodServiceDefinition> services;
+    private MapProperty<String, PodVolumeDefinition> volumes;
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/PodDistributionExtension.java]: final?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/tasks/CreatePodYamlTask.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    @TaskAction
+    final void createPodYaml() throws IOException {
+        validatePodYaml();
+        OBJECT_MAPPER.writeValue(getPodYamlFile().getAsFile().get(), ImmutableMap.of(
+                "services",
+                OBJECT_MAPPER.convertValue(this.serviceDefinitions.get(), new TypeReference<Map<String, Object>>() {}),
+                "volumes",
+                OBJECT_MAPPER.convertValue(this.volumeDefinitions.get(), new TypeReference<Map<String, Object>>() {})));
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/tasks/CreatePodYamlTask.java]: Can we do without these Object mappers convertValues?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    private Property<String> mainClass;
+    private Property<String> javaHome;
+    private Property<Boolean> addJava8GcLogging;
+    private Property<Boolean> enableManifestClasspath;
+    private Property<GcProfile> gc;
+    private ListProperty<String> args;
+    private ListProperty<String> checkArgs;
+    private ListProperty<String> defaultJvmOpts;
+    private ListProperty<String> excludeFromVar;
+    private MapProperty<String, String> env;
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.java]: All final pls

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+                            GFileUtils.writeFile(cleanedText, task.getWindowsScript());
+                        }
+                    });
+                });
+
+        // HACKHACK ensure runtimeClassPath exists
+        project.afterEvaluate(p -> startScripts.configure(task -> {
+            JavaPluginConvention javaPlugin = project.getConvention().findPlugin(JavaPluginConvention.class);
+            task.setClasspath(manifestClassPathTask.get().getOutputs().getFiles().plus(
+                    javaPlugin.getSourceSets().getByName("main").getRuntimeClasspath()));
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java:134]: ??? We already called `distributionExtension.setProductDependenciesConfig()` with runtimeClasspath - could we flesh this comment out to explain why this stuff is necessary? (is it because CreateStartScripts isn't our class and doesn't accept providers??)

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        }));
+
+        TaskProvider<LaunchConfigTask> launchConfigTask = project.getTasks().register(
+                "createLaunchConfig", LaunchConfigTask.class, task -> {
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java:138]: :+1: big fan of doing the configuration straight away in the plugin rather than hiding it in task classes!

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+        TaskProvider<JavaExec> runTask = project.getTasks().register("run", JavaExec.class, task -> {
+            TaskProvider<Jar> jarTaskProvider = project.getTasks().named("jar", Jar.class);
+            task.setGroup(JavaServiceDistributionPlugin.GROUP_NAME);
+            task.setDescription("Runs the specified project using configured mainClass and with default args.");
+
+            task.dependsOn(jarTaskProvider);
+        });
+
+        // HACKHACK
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java]: I'm concerned that 'HACKHACK' comments will just stay in the codebase forever as people will be scared to touch them.

Could we turn these into precise justifications of why the afterEvaluate is necessary and what criteria could result in it's removal? (then when someone reads this code 2 years from now they might realise the criteria are met and just delete the block)?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.JavaExec;
+import org.gradle.api.tasks.TaskProvider;
+import org.gradle.api.tasks.application.CreateStartScripts;
+import org.gradle.api.tasks.bundling.Compression;
+import org.gradle.api.tasks.bundling.Jar;
+import org.gradle.api.tasks.bundling.Tar;
+import org.gradle.util.GFileUtils;
+
+public class JavaServiceDistributionPlugin implements Plugin<Project> {
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java]: FYI I think plugin classes can be marked as `public final` - it's only the extensions and tasks that Gradle needs to be non-final (as it creates dynamic proxies for them).

We do this in [gradle-conjure](https://github.com/palantir/gradle-conjure/blob/develop/gradle-conjure/src/main/java/com/palantir/gradle/conjure/ConjurePlugin.java#L45)

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateCheckScriptTask.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @TaskAction
+    final void createInitScript() {
+        if (!checkArgs.get().isEmpty()) {
+            EmitFiles.replaceVars(
+                    JavaServiceDistributionPlugin.class.getResourceAsStream("/check.sh"),
+                    getOutputFile().getAsFile().get().toPath(),
+                    ImmutableMap.of("@serviceName@", serviceName.get(),
+                     "@checkArgs@", Joiner.on(" ").join(checkArgs.get())))
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateCheckScriptTask.java]: Formatting?
```suggestion
                    ImmutableMap.of(
                            "@serviceName@", serviceName.get(),
                            "@checkArgs@", Joiner.on(" ").join(checkArgs.get())))
```

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateCheckScriptTask.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+public class CreateCheckScriptTask extends DefaultTask {
+    private Property<String> serviceName = getProject().getObjects().property(String.class);
+    private ListProperty<String> checkArgs = getProject().getObjects().listProperty(String.class);
+    private RegularFileProperty outputFile = getProject().getObjects().fileProperty();
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateCheckScriptTask.java]: final pls (we should realyl have an errorprone check for this!!!

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateInitScriptTask.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+public class CreateInitScriptTask extends DefaultTask {
+    private Property<String> serviceName = getProject().getObjects().property(String.class);
+    private RegularFileProperty outputFile = getProject().getObjects().fileProperty();
+
+    public CreateInitScriptTask() {
+        outputFile.set(new File(getProject().getBuildDir(), "scripts/init.sh"));
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateInitScriptTask.java]: I just tried using `getProject().getLayout().getBuildDirectory()` to tolerate cases where the user opts to change the `build` directory, but it makes everything way less readable.

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateStartScriptsTask.groovy [hunk 16]
```diff
@@ -1,64 +0,0 @@
...
-
-            doLast {
-                Jar manifestClasspathJarTask = project.tasks.getByName('manifestClasspathJar')
-                if (!manifestClasspathJarTask) {
-                    throw new GradleException("Required task not found: manifestClasspathJar")
-                }
-
-                if (isEnableManifestClasspath) {
-                    // Replace standard classpath with pathing jar in order to circumnavigate length limits:
-                    // https://issues.gradle.org/browse/GRADLE-2992
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/CreateStartScriptsTask.groovy:49]: Lol I just checked whether this is still open...

> We haven't heard back from you after our inquiry from November 15th. We are closing this issue now. Please create an issue on GitHub if you still feel passionate about getting it resolved.

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.java [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+    @Input
+    public final Property<String> getServiceName() {
+        return serviceName;
+    }
+
+    public final Property<GcProfile> getGc() {
+        return gc;
+    }
+
+    // HACKHACK Property<GcProfile> failed to serialise
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.java:106]: Is this still true after adding Serializabel to GcProfile? Perhaps it needs a serialVersionUid?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/ConfigTarTask.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+                    throw new RuntimeException(e);
+                }
+            }));
+
+            task.from(new File(project.getProjectDir(), "deployment"));
+            task.from(new File(project.getBuildDir(), "deployment"));
+        });
+
+        // HACKHACK
+        // TODO(forozco): make this lazy since into does not support providers, but does support callable
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/ConfigTarTask.java]: :+1: comments looks good, 'HACKHACK' is unnecessary ;)

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.java [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+    private Property<String> serviceGroup = getProject().getObjects().property(String.class);
+    private Property<ProductType> productType = getProject().getObjects().property(ProductType.class);
+
+    private MapProperty<String, Object> manifestExtensions = getProject().getObjects()
+            .mapProperty(String.class, Object.class);
+    private ListProperty<ProductDependency> productDependencies = getProject().getObjects()
+            .listProperty(ProductDependency.class);
+    private SetProperty<ProductId> ignoredProductIds = getProject().getObjects().setProperty(ProductId.class).empty();
+
+    private RegularFileProperty manifestFile = getProject().getObjects().fileProperty();
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.java]: So many finals pls

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.java [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    public final void setProductDependenciesConfig(Configuration productDependenciesConfig) {
+        this.productDependenciesConfig = productDependenciesConfig;
+    }
+
+    @Input
+    final String getProjectVersion() {
+        String stringVersion = String.valueOf(getProject().getVersion());
+        if (!SlsProductVersions.isValidVersion(stringVersion)) {
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.java]: I know all this validation logic lived in the `getProjectVersion` method before, but it feels a bit weird because Gradle might well run this method to determine whether the task is up to date or not and I think the failures are probably just annoying in that case.

Could we move it to a `validateProjectVersion` method or something that only gets run in the actual @TaskAction?

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy [hunk 21]
```diff
@@ -46,87 +44,68 @@ class BaseDistributionExtensionTest extends Specification {
...
+        ext.setServiceGroup("bar")
+        ext.serviceGroup.get() == "bar"
     }
 
-    def "productDependencies from maven coordinate with all fields and no recommended version"() {
+    // TODO(forozco): verify that enums work correctly
+//    def 'productType only accepts valid values'() {
+//        when:
+//        def ext = new BaseDistributionExtension(project, project.getObjects())
+//        ext.setProductType("foobar")
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy]: probably safe to delete this test because nobody is actually expected to call `setProductType` right? We could possible even make it package private?

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy [hunk 22]
```diff
@@ -46,87 +44,68 @@ class BaseDistributionExtensionTest extends Specification {
...
+//
+//        then:
+//        def ex = thrown IllegalArgumentException
+//        ex.message == "Invalid product type 'foobar' specified; supported types: [service.v1, daemon.v1, asset.v1, pod.v1]."
+//    }
+
+    def "productDependencies from closure"() {
         when:
-        def ext = new BaseDistributionExtension(null)
-        ext.productDependency("group:name:1.2.3:classifier@tgz")
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy]: Does this style of setting productDependencies no longer work?

**Review by iamdanfox — COMMENTED** [review]
Heroic work @ferozco - I've read through all the files as carefully as I could, looks actually very unscary despite the size.

Before merging, I think it would be good to add at least one test that proves sls-packaging still works on some minimum version of gradle (I think providers worked fine in 4.10.2?) and then document this minimum version in the README!

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy (hunk 2 above)
Comment (ferozco) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy]: They will be broken. Users will no longer have to specify that they are detecting constraints for certain product dependencies. I'm planning on getting roomba to remove these entries as part of the roll out

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/tasks/CreatePodYamlTask.java (hunk 7 above)
Comment (ferozco) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/tasks/CreatePodYamlTask.java]: Since the types are heterogeneous, It was between doing a convert values or creating an immutable for this single usage. Happy to switch over to immutable if you think its cleaner

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy (hunk 22 above)
Comment (ferozco) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy]: that's correct, you have to use the closure

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.java (hunk 17 above)
Comment (ferozco) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.java:106]: yup, it might be because its just an interface?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+public final class JavaServiceDistributionPlugin implements Plugin<Project> {
+    private static final String GO_JAVA_LAUNCHER_BINARIES = "goJavaLauncherBinaries";
+    private static final String GO_JAVA_LAUNCHER = "com.palantir.launching:go-java-launcher:1.5.1";
+    private static final String GO_INIT = "com.palantir.launching:go-init:1.5.1";
+    public static final String GROUP_NAME = "Distribution";
+    private static final String SLS_CONFIGURATION_NAME = "sls";
+
+    @SuppressWarnings("checkstyle:methodlength")
+    public void apply(Project project) {
+        if (project.getPlugins().hasPlugin(AssetDistributionPlugin.class)) {
```
Comment (robert3005) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java:53]: this should be withPlugin since this will only ever throw if this plugin is applied after the asset distribution plugin

**Comment by ferozco** [comment]
After testing internal usage of the new plugin, I've found a case where some validation was being done on the product dependency configuration. In particular, the internal tool assumed that there was a corresponding entry for all product dependencies. This assumption is broken by automatically discovering all dependencies. I'm going to go ahead and revert the changes to product dependencies in this PR, and in a follow up expose a task produces all resolved product dependencies

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.java [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+    public final void productDependency(
+            String dependencyGroup, String dependencyName, String minVersion, String maxVersion) {
+        productDependency(dependencyGroup, dependencyName, minVersion, maxVersion,  null);
+    }
+
+    public final void productDependency(
+            String dependencyGroup,
+            String dependencyName,
+            String minVersion,
+            String maxVersion, String recommendedVersion) {
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.java:159]: newline

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.java [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+    public final Configuration getProductDependenciesConfig() {
+        return productDependenciesConfig;
+    }
+
+    public final void setProductDependenciesConfig(Configuration productDependenciesConfig) {
+        this.productDependenciesConfig = productDependenciesConfig;
+    }
+
+    static String generateMaxVersion(String minimumVersion) {
+        return String.format("%s.x.x", Iterables.getFirst(Splitter.on(".").split(minimumVersion), null));
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.java:214]: maybe throw here if you couldn't get the first element

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/PodDistributionPlugin.java [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+
+        TaskProvider<CreateManifestTask> manifest = CreateManifestTask.createManifestTask(
+                project, distributionExtension);
+
+        TaskProvider<CreatePodYamlTask> podYaml = project.getTasks().register(
+                "createPodYaml", CreatePodYamlTask.class, task -> {
+                    task.setGroup(PodDistributionPlugin.GROUP_NAME);
+                    task.setDescription("Generates a simple yaml file describing a pods constituent services.");
+                });
+        project.afterEvaluate(p -> podYaml.configure(task -> {
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/pod/PodDistributionPlugin.java:56]: don't _think_ you need afterEvluate here since the configure is lazy (until the task is actually initialized)

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java (hunk 23 above)
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java:53]: I think this is fine since that would imply that this only runs after the 2nd plugin got applied.
And it most likely will fail to register a duplicate extension before it even gets to that, which means you get less good errors.

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 27]
```diff
@@ -0,0 +1,217 @@
...
+
+        // Create configuration to load executable dependencies
+        project.getConfigurations().maybeCreate(GO_JAVA_LAUNCHER_BINARIES);
+        project.getDependencies().add(GO_JAVA_LAUNCHER_BINARIES, GO_JAVA_LAUNCHER);
+        project.getDependencies().add(GO_JAVA_LAUNCHER_BINARIES, GO_INIT);
+
+        TaskProvider<CopyLauncherBinariesTask> copyLauncherBinaries = project.getTasks()
+                .register("copyLauncherBinaries", CopyLauncherBinariesTask.class);
+
+        TaskProvider<Jar> manifestClassPathTask = project.getTasks().register(
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java:76]: wasn't obvious this is a jar task, can we call it `manifestClassPathJarTask`?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java (hunk 9 above)
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java:134]: That one's for determining what your product dependencies are, this one's determining what will end up in the lib/ folder.
I do note that there's a difference in behaviour here:
* before, we'd get the output of the plain `jar` task
* now, we get the output of the `manifestClassPathJar` task

Don't think it'd be a problem to do the latter, but I'd rather revert to the previous behaviour so we can better isolate.

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java [hunk 28]
```diff
@@ -0,0 +1,217 @@
...
+        TaskProvider<Tar> configTar = ConfigTarTask.createConfigTarTask(project, distributionExtension);
+        configTar.configure(task -> task.dependsOn(manifest));
+
+
+        TaskProvider<JavaExec> runTask = project.getTasks().register("run", JavaExec.class, task -> {
+            Task jarTaskProvider = project.getTasks().getByName("jar");
+            task.setGroup(JavaServiceDistributionPlugin.GROUP_NAME);
+            task.setDescription("Runs the specified project using configured mainClass and with default args.");
+
+            task.dependsOn(jarTaskProvider);
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java]: can just say `dependsOn("jar")`

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/DistTarTask.groovy [hunk 29]
```diff
@@ -15,22 +15,12 @@
...
-        project.tasks.create(taskName, Tar) { p ->
-            p.group = JavaServiceDistributionPlugin.GROUP_NAME
-            p.description = "Creates a compressed, gzipped tar file that contains required runtime resources."
-            // Set compression in constructor so that task output has the right name from the start.
-            p.compression = Compression.GZIP
-            p.extension = 'sls.tgz'
-        }
-    }
 
     static void configure(Tar distTar, Project project, String serviceName, List<String> excludeFromVar, boolean isEnableManifestClasspath) {
```
Comment (dansanduleac) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/DistTarTask.groovy:25]: let's migrate the rest of this file in another PR

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java (hunk 9 above)
Comment (dansanduleac) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.java:134]: Original logic: https://github.com/palantir/sls-packaging/pull/385/files#diff-b6d65383586925984285f689eb47b120L31

**Comment by ferozco** [comment]
👍
